TFA confirmation

IceDragon200 opened the pull request.

IceDragon200 committed:
Working on TFA confirmation

IceDragon200 committed:
Some fixes to the tfa controller

IceDragon200 committed:
More work on TFA controller including tests

IceDragon200 committed:
Moved controller implementation out of quote block

IceDragon200 committed:
Make tfa controller functions overridable

archseer commented on lib/passport/controllers/two_factor_auth_controller.ex:
```diff
@@ -0,0 +1,126 @@
+  def create(_controller, conn, params) do
+    case Passport.Sessions.authenticate_entity(params["email"], params["password"]) do
+      {:ok, %{tfa_enabled: true} = entity} ->
+        send_unauthorized(conn, reason: "TFA already enabled")
```
Not sure if "already done" should return an unauthorized error

IceDragon200 replied:
409 Conflict maybe, though this route would probably be changed to return the tfa secret key instead of a confirmation token, so it would be a 403 instead

archseer commented on lib/passport/controllers/two_factor_auth_controller.ex:
```diff
@@ -0,0 +1,126 @@
+
+  def confirm(controller, conn, params) do
+    case Passport.find_by_tfa_confirmation_token(controller.two_factor_auth_model(), conn.path_params["token"]) do
+      nil -> Passport.APIHelper.send_not_found(conn)
```
Could probably simplify by using cond

archseer commented on lib/passport/modules/two_factor_auth.ex:
```diff
@@ -4,20 +4,29 @@ defmodule Passport.TwoFactorAuth do
 
   defmacro schema_fields do
     quote do
+      field :tfa_confirmation_token, :string
```
?

archseer replied:
2fa has no confirmation token, it should work just like entering 2fa normally --- user gives us the passphrase and we validate with potp

IceDragon200 committed:
Changed tfa confirmation behaviour

Instead do `POST /confirm/tfa`, which will reset the tfa secret key and provides a new one (note that the user will have TFA disabled until they confirm that the new code works)

Might consider just leaving tfa_enabled if it's already enabled, forcing the user to authenticate using the new code, but this may end up locking the account if they somehow manage to misplace the new secret code

Afterwards do a `PUT /confirm/tfa` using the same authentication from the POST with the addition of the new OTP code to confirm

archseer commented on test/passport/controllers/two_factor_auth_controller_test.exs:
```diff
@@ -0,0 +1,89 @@
+  use Passport.Support.Web.ConnCase
+
+  describe "POST /account/confirm/tfa" do
+    test "can reset tfa_otp_secret_key", %{conn: conn} do
```
What?

archseer replied:
Why would confirming reset the key

archseer replied:
Oh this is for starting a confirmation? What happens if I never trigger it? We don't need to manually trigger a mail send

IceDragon200 replied:
This doesn't send any mail at all, and you don't ever have to trigger it, if you already know the secret key

What this route does: it regenerates the secret key given the user's credentials (and current otp if tfa is enabled), it then disables tfa so the confirm can work correctly

So it's 2 steps: POST then PUT

IceDragon200 committed:
Moved reset tfa route

IceDragon200 committed:
Fixed tfa tests

IceDragon200 committed:
Allow create and delete in sessions controller to be overriden

IceDragon200 committed:
Allow overriding session error handling

IceDragon200 committed:
Fixed issue with session controller delete action

IceDragon200 committed:
Replaced test tfa_otp_secret_key index

IceDragon200 committed:
Add some additional api helpers

IceDragon200 committed:
Modified Sessions and SessionController

Added TFA mandatory setup

IceDragon200 committed:
Fixed authentication not handling nil secret keys

IceDragon200 committed:
Split authenticate_entity into basic_authenticate_entity

Which only checks the email and password

archseer reviewed (approved):
LGTM, passes review now; great work! 🎊
